File: docs/examples/restful_api_client.py
#!/usr/bin/env python3
"""Example RESTful API client for weeb-cli.

This script demonstrates how to interact with the weeb-cli RESTful API server.
"""
import requests
from typing import List, Dict, Optional


class WeebCLIClient:
    """Client for weeb-cli RESTful API."""
    
    def __init__(self, base_url: str = "http://localhost:8080"):
        """Initialize client with base URL.
        
        Args:
            base_url: Base URL of the weeb-cli RESTful API server
        """
        self.base_url = base_url.rstrip("/")
        self.session = requests.Session()
    
    def health_check(self) -> Dict:
        """Check if the server is running.
        
        Returns:
            Health status dictionary
        """
        response = self.session.get(f"{self.base_url}/health")
        response.raise_for_status()
        return response.json()
    
    def list_providers(self) -> Dict:
        """List all available providers.
        
        Returns:
            Dictionary with providers list
        """
        response = self.session.get(f"{self.base_url}/api/providers")
        response.raise_for_status()
        return response.json()
    
    def search(self, query: str, provider: Optional[str] = None) -> List[Dict]:
        """Search for anime.
        
        Args:
            query: Search query
            provider: Provider name (optional)
        
        Returns:
            List of anime results
        """
        params = {"q": query}
        if provider:
            params["provider"] = provider
        
        response = self.session.get(f"{self.base_url}/api/search", params=params)
        response.raise_for_status()
        data = response.json()
        return data.get("results", [])
    
    def get_anime_details(self, anime_id: str, provider: Optional[str] = None) -> Dict:
        """Get anime details.
        
        Args:
            anime_id: Anime ID
            provider: Provider name (optional)
        
        Returns:
            Anime details dictionary
        """
        params = {}
        if provider:
            params["provider"] = provider
        
        response = self.session.get(
            f"{self.base_url}/api/anime/{anime_id}",
            params=params
        )
        response.raise_for_status()
        data = response.json()
        return data.get("anime", {})
    
    def get_episodes(
        self,
        anime_id: str,
        season: Optional[int] = None,
        provider: Optional[str] = None
    ) -> List[Dict]:
        """Get anime episodes.
        
        Args:
            anime_id: Anime ID
            season: Season number (optional)
            provider: Provider name (optional)
        
        Returns:
            List of episodes
        """
        params = {}
        if season:
            params["season"] = season
        if provider:
            params["provider"] = provider
        
        response = self.session.get(
            f"{self.base_url}/api/anime/{anime_id}/episodes",
            params=params
        )
        response.raise_for_status()
        data = response.json()
        return data.get("episodes", [])
    
    def get_streams(
        self,
        anime_id: str,
        episode_id: str,
        provider: Optional[str] = None,
        sort: str = "desc"
    ) -> List[Dict]:
        """Get episode streams.
        
        Args:
            anime_id: Anime ID
            episode_id: Episode ID
            provider: Provider name (optional)
            sort: Sort order ('asc' or 'desc')
        
        Returns:
            List of streams
        """
        params = {"sort": sort}
        if provider:
            params["provider"] = provider
        
        response = self.session.get(
            f"{self.base_url}/api/anime/{anime_id}/episodes/{episode_id}/streams",
            params=params
        )
        response.raise_for_status()
        data = response.json()
        return data.get("streams", [])


def main():
    """Example usage of the WeebCLIClient."""
    # Initialize client
    client = WeebCLIClient("http://localhost:8080")
    
    # Health check
    print("=== Health Check ===")
    health = client.health_check()
    print(f"Status: {health['status']}")
    print(f"Providers: {', '.join(health['providers'])}")
    print()
    
    # List providers
    print("=== Available Providers ===")
    providers_data = client.list_providers()
    for provider in providers_data["providers"]:
        print(f"- {provider['name']} ({provider['lang']}/{provider['region']})")
    print()
    
    # Search anime
    print("=== Search Results ===")
    query = "naruto"
    results = client.search(query, provider="animecix")
    print(f"Found {len(results)} results for '{query}':")
    for i, anime in enumerate(results[:5], 1):
        print(f"{i}. {anime['title']} ({anime['year']}) - ID: {anime['id']}")
    print()
    
    if not results:
        print("No results found. Exiting.")
        return
    
    # Get anime details
    print("=== Anime Details ===")
    anime_id = results[0]["id"]
    details = client.get_anime_details(anime_id, provider="animecix")
    print(f"Title: {details['title']}")
    print(f"Type: {details['type']}")
    print(f"Year: {details['year']}")
    print(f"Status: {details.get('status', 'N/A')}")
    if details.get('genres'):
        print(f"Genres: {', '.join(details['genres'])}")
    print()
    
    # Get episodes
    print("=== Episodes (Season 1) ===")
    episodes = client.get_episodes(anime_id, season=1, provider="animecix")
    print(f"Found {len(episodes)} episodes:")
    for ep in episodes[:5]:
        title = ep.get('title') or f"Episode {ep['number']}"
        print(f"- S{ep['season']:02d}E{ep['number']:02d}: {title}")
    print()
    
    if not episodes:
        print("No episodes found. Exiting.")
        return
    
    # Get streams for first episode
    print("=== Streams (First Episode) ===")
    episode_id = episodes[0]["id"]
    streams = client.get_streams(anime_id, episode_id, provider="animecix")
    print(f"Found {len(streams)} streams:")
    for stream in streams:
        print(f"- Quality: {stream['quality']}")
        print(f"  Server: {stream['server']}")
        print(f"  URL: {stream['url'][:60]}...")
        if stream.get('subtitles'):
            print(f"  Subtitles: {stream['subtitles']}")
    print()


if __name__ == "__main__":
    try:
        main()
    except requests.exceptions.ConnectionError:
        print("Error: Could not connect to weeb-cli RESTful API server.")
        print("Make sure the server is running: weeb-cli serve restful")
    except requests.exceptions.HTTPError as e:
        print(f"HTTP Error: {e}")
    except KeyboardInterrupt:
        print("\nInterrupted by user.")

File: docs/examples/restful_api_client.js
#!/usr/bin/env node
/**
 * Example RESTful API client for weeb-cli (Node.js/JavaScript)
 * 
 * This script demonstrates how to interact with the weeb-cli RESTful API server.
 * 
 * Usage:
 *   node restful_api_client.js
 */

const BASE_URL = process.env.WEEB_CLI_URL || 'http://localhost:8080';

/**
 * WeebCLI API Client
 */
class WeebCLIClient {
  constructor(baseUrl = BASE_URL) {
    this.baseUrl = baseUrl.replace(/\/$/, '');
  }

  /**
   * Make HTTP GET request
   */
  async get(endpoint, params = {}) {
    const url = new URL(`${this.baseUrl}${endpoint}`);
    Object.keys(params).forEach(key => {
      if (params[key] !== null && params[key] !== undefined) {
        url.searchParams.append(key, params[key]);
      }
    });

    const response = await fetch(url.toString());
    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Health check
   */
  async healthCheck() {
    return this.get('/health');
  }

  /**
   * List all providers
   */
  async listProviders() {
    return this.get('/api/providers');
  }

  /**
   * Search anime
   */
  async search(query, provider = null) {
    const data = await this.get('/api/search', { q: query, provider });
    return data.results || [];
  }

  /**
   * Get anime details
   */
  async getAnimeDetails(animeId, provider = null) {
    const data = await this.get(`/api/anime/${animeId}`, { provider });
    return data.anime || {};
  }

  /**
   * Get episodes
   */
  async getEpisodes(animeId, season = null, provider = null) {
    const data = await this.get(`/api/anime/${animeId}/episodes`, { season, provider });
    return data.episodes || [];
  }

  /**
   * Get streams
   */
  async getStreams(animeId, episodeId, provider = null, sort = 'desc') {
    const data = await this.get(
      `/api/anime/${animeId}/episodes/${episodeId}/streams`,
      { provider, sort }
    );
    return data.streams || [];
  }
}

/**
 * Main example function
 */
async function main() {
  const client = new WeebCLIClient();

  try {
    // Health check
    console.log('=== Health Check ===');
    const health = await client.healthCheck();
    console.log(`Status: ${health.status}`);
    console.log(`Providers: ${health.providers.join(', ')}`);
    console.log();

    // List providers
    console.log('=== Available Providers ===');
    const providersData = await client.listProviders();
    providersData.providers.forEach(provider => {
      console.log(`- ${provider.name} (${provider.lang}/${provider.region})`);
    });
    console.log();

    // Search anime
    console.log('=== Search Results ===');
    const query = 'naruto';
    const results = await client.search(query, 'animecix');
    console.log(`Found ${results.length} results for '${query}':`);
    results.slice(0, 5).forEach((anime, i) => {
      console.log(`${i + 1}. ${anime.title} (${anime.year}) - ID: ${anime.id}`);
    });
    console.log();

    if (results.length === 0) {
      console.log('No results found. Exiting.');
      return;
    }

    // Get anime details
    console.log('=== Anime Details ===');
    const animeId = results[0].id;
    const details = await client.getAnimeDetails(animeId, 'animecix');
    console.log(`Title: ${details.title}`);
    console.log(`Type: ${details.type}`);
    console.log(`Year: ${details.year}`);
    console.log(`Status: ${details.status || 'N/A'}`);
    if (details.genres && details.genres.length > 0) {
      console.log(`Genres: ${details.genres.join(', ')}`);
    }
    console.log();

    // Get episodes
    console.log('=== Episodes (Season 1) ===');
    const episodes = await client.getEpisodes(animeId, 1, 'animecix');
    console.log(`Found ${episodes.length} episodes:`);
    episodes.slice(0, 5).forEach(ep => {
      const title = ep.title || `Episode ${ep.number}`;
      const season = String(ep.season).padStart(2, '0');
      const number = String(ep.number).padStart(2, '0');
      console.log(`- S${season}E${number}: ${title}`);
    });
    console.log();

    if (episodes.length === 0) {
      console.log('No episodes found. Exiting.');
      return;
    }

    // Get streams for first episode
    console.log('=== Streams (First Episode) ===');
    const episodeId = episodes[0].id;
    const streams = await client.getStreams(animeId, episodeId, 'animecix');
    console.log(`Found ${streams.length} streams:`);
    streams.forEach(stream => {
      console.log(`- Quality: ${stream.quality}`);
      console.log(`  Server: ${stream.server}`);
      console.log(`  URL: ${stream.url.substring(0, 60)}...`);
      if (stream.subtitles) {
        console.log(`  Subtitles: ${stream.subtitles}`);
      }
    });
    console.log();

  } catch (error) {
    if (error.code === 'ECONNREFUSED') {
      console.error('Error: Could not connect to weeb-cli RESTful API server.');
      console.error('Make sure the server is running: weeb-cli serve restful');
    } else {
      console.error(`Error: ${error.message}`);
    }
    process.exit(1);
  }
}

// Run main function
if (require.main === module) {
  main().catch(error => {
    console.error('Unhandled error:', error);
    process.exit(1);
  });
}

module.exports = { WeebCLIClient };
